samza-core/src/main/java/org/apache/samza/clustermanager/ContainerManager.java

```diff
@@ -109,17 +144,26 @@ void handleContainerStop(String processorId, String containerId, String preferre
 
   /**
    * Handle the container launch failure for active containers and standby (if enabled).
-   * Case 1. When standby is enabled, refer to {@link StandbyContainerManager#handleContainerLaunchFail} to check behavior
-   * Case 2. When standby is disabled the allocator issues a request for ANY_HOST resources
    *
-   * @param processorId logical id of the container
+   * Case 1. If this launch was issued due to an existing container placement action update the metadata to report failure and issue
+   *         a request for source host where the container was last seen and mark the container placement failed
+   * Case 2. When standby is enabled, refer to {@link StandbyContainerManager#handleContainerLaunchFail} to check behavior
+   * Case 3. When standby is disabled the allocator issues a request for ANY_HOST resources
+   *
+   * @param processorId logical id of the container eg 1,2,3
    * @param containerId last known id of the container deployed
    * @param preferredHost host on which container is requested to be deployed
    * @param containerAllocator allocator for requesting resources
    */
   void handleContainerLaunchFail(String processorId, String containerId, String preferredHost,
       ContainerAllocator containerAllocator) {
-    if (processorId != null && standbyContainerManager.isPresent()) {
+    if (processorId != null && getControlActionMetadata(processorId).isPresent()) {
+      ContainerPlacementMetadata metaData = getControlActionMetadata(processorId).get();
```

Comment:
See above comment on wrapper function isControlActionRequest(), 
also, can move  Optional/<ContainerPlacementMetadata> metaData = getControlActionMetadata(processorId) outside the if